# 54. flatten Thunk

Suppose we have a Callback type

```js
type Callback =   (error: Error, result: any | Thunk) => void
```

A Thunk is a function that take a Callback as parameter

```js
type Thunk = (callback: Callback) => void
```

Like following three thunks

```js
const func1 = (cb) => {
  setTimeout(() => cb(null, 'ok'), 10)
}
const func2 = (cb) => {
  setTimeout(() => cb(null, func1), 10)
}
const func3 = (cb) => {
  setTimeout(() => cb(null, func2), 10)
}
```

in above example, three functions are kind of chained up, func3 â†’ func2 â†’ func1, but it don't work without some glue.

OK, now you are asked to implement a `flattenThunk()` which glue them up and returns a new thunk.

```js
flattenThunk(func3)((error, data) => {   console.log(data) // 'ok'})
```

*note*

Once error occurs, the rest uncalled functions should be skipped

## Solution

```js
/**
 * @param {Thunk} thunk ðŸ’¡func3
 *      const func1 = (cb) => {
          setTimeout(() => cb(null, 'ok'), 10)
        }
        const func2 = (cb) => {
          setTimeout(() => cb(null, func1), 10)
        }
        const func3 = (cb) => {
          setTimeout(() => cb(null, func2), 10)
        }
 * @return {Thunk}
 */
function flattenThunk(thunk) {
  // ðŸ’¡callback: (error, data) => { console.log(data) }
  return function(callback) {

    const handleThunk = (currentThunk) => {
      currentThunk((error, result) => {
        if (error) {
          // If there's an error, stop the chain and call the callback with the error
          // ðŸ’¡console.log(error)
          callback(error);
        } else if (typeof result === 'function') {
          // If the result is a thunk, call it
          // ðŸ’¡result: func2
          // ðŸ’¡result: func1
          handleThunk(result);
        } else {
          // If the result is a value, return it via the callback
          // ðŸ’¡console.log("ok")
          callback(undefined, result);
        }
      });
    };
	
    // ðŸ’¡thunk: func3
    handleThunk(thunk);
  };
}

```

